### ISSUE TITLE
Ensure all Coherence messages can be translated in the user's project (was Looking for Translations)

### ISSUE THREAD

**Issue by smpallen99** [smpallen99/coherence#185]
Now that we have gettext support, we need some translations. Please open a separate issue for each translation you are looking for, or want to contribute.

**Comment by tmbb** [comment]
Not opening a new issue because it's more of a meta point. How are we supposed to add new translations? The logical place seems to be a in `gettext/<locale>/LC_MESSAGES/coherence.po` directories inside the coherence project templates, so that the file would be added to the user's project. The problem with this is that a user who installed coherence would get possibly dozens of unwanted locales, but as each locale will have it's own directory it'd be easy for the user to delete the unwanted ones.

A more advance approach would be adding a `coherence.localize` mix task which would add the relevant locale directory to the user project, but that would complicate the code for relatively meager benefit.

I'd suggest you placed a dummy locale directory (an `en` locale, for example) so that the translators have a template to work with.

**Comment by agustif** [comment]
I'm waiting to see an easy way (like what branch do I fork, gettext , phx-1.3?) to translate to spanish and catalan!

**Comment by tmbb** [comment]
I've just read the full gettext documentation and it turns out what I propose seems to be impossible... Apparently gettext translations for a certain app must be stored inside that app's directory (?). This means that coherence's translations must be bundled with the app and that the user can't add a new locale or override the "official" translations. If I'm understanding this right, it's a serious design fault in gettext.

**Comment by tmbb** [comment]
About two months ago user **bitboxer** (presumably @bitboxer on github) asked some questions on the [elixir forum](https://elixirforum.com/t/gettext-in-a-hex-package/3447/9) about internationalizing coherence.

At the time, @josevalim [offered to help](https://elixirforum.com/t/gettext-in-a-hex-package/3447/10) the coherence project with implementing the gettext backend correctly. I don't think anything came out of it at the time.

Now that all the dgettext calls are in place and we have several volunteer translators, I've [asked](https://elixirforum.com/t/gettext-in-a-hex-package/3447/12) @josevalim to come here and give his opinion on how to best internationalize coherence in a way that allows the user to customize translations and add locales.

**Comment by bitboxer** [comment]
[This] is the PR that introduced a global per processes locale for gettext. That is the part I asked about. Now you can set a global locale and every hex package can use that and doesn't have to be notified about the language separately. Adding translations that make it possible for users to customize them is a nother step that is missing right now I think.

**Comment by tmbb** [comment]
@bitboxer 

> Adding translations that make it possible for users to customize them is a nother step that is missing right now I think.

Yes, it's still missing, but that's what @josevalim volunteered to help us with, right?

**Comment by smpallen99** [comment]
Wow, I created this issue  last night and woke up to all this discussion. Awesome, I guess this is a hot topic 👍. Before we figure out how we are going to do this in Coherence it looks like to need to understand technically what our options are. 

One other option that wan't discussed above might be to create packages for each translation and have users add then ones they want to their deps. I'm not sure this is the way to go, but I'm just throwing it out there for discussion.

I look forward to hearing back from José.

**Comment by josevalim** [comment]
>  If I'm understanding this right, it's a serious design fault in gettext (I'm probably missing something, though)

Coherence should use your application gettext backend. So there should be a configuration in coherence that says: "here is the gettext module I want you to use". And then coherence will call gettext like this:

    Gettext.dgettext gettext_backend_from_config(), "coherence", "Hello world"

As a convenience, coherence can provide generators that copy the `coherence.pot` file to the user application. With the template, users can translate it to different languages in their own backend.

The downside of this solution is that coherence won't be able to use the gettext helpers to extract translations. Although users of coherence and gettext should see no loss of functionality.

**Comment by tmbb** [comment]
This is a good solution, and it seems easy to implement. On the other hand, I've noticed that most of the user facing strings will end up in the user's application (they are inside the install.coherence/ dir) the only exception are the strings in the plugs/ dir.

Maybe we could use the macros for those files (because inside those files we can import our app's Gettext backend - I'm doing this now with good results). This means we can generate the .pot file the usual way and gove the user a better user experience when editing the templates or views.

We can't do this with the plugs, so for the plugs we could use @josevalim's suggestion.

So, to summarize: macros for tenplates and views, functions with configurable backend for plugs. What do you think @smpallen99?

**Comment by smpallen99** [comment]
I'm not sure yet @tmbb. I'm in the middle of getting the phx-1.3 branch ready to support both the legacy and phx project structures. Hoping to have that complete today. Then I'll swing back and look at gettext stuff. just don't want to do the context switch yet. I'd also like to try things out to get more familiar with gettext.

**Comment by smpallen99** [comment]
@ttmb would hold off working on the gettext branch. Once I get my stuff back into the phx-1.3 branch we can work there. When we have a solution for gettext, we can merge phx-1.3 into master.

**Comment by tmbb** [comment]
Just something I thought, feel free tor revisit it later when you look into the gettext part more closely: if we're going with the approach of having both macros and functions we should have a "coherence" namespace for macros (.pot file can be regenerated) and a "coherence.plug" namespace (can't be regenerated)

**Comment by josevalim** [comment]
I am not sure what you mean by "namespace" but Gettext allows you to mixed generated translations with static ones. And I would advise leaking implementation concerns such as "coherence.plug" into users of the library.

In any case, I don't think the macro usage would be possible in any case, as coherence has to be dynamic by definition.

**Comment by tmbb** [comment]
> I am not sure what you mean by "namespace"

@josevalim I should have said "domain" instead of "namespace", sorry. If gettext can mix static and dynamic translation in the same file, the great! No need for separate domains then.

> In any case, I don't think the macro usage would be possible in any case, as coherence has to be dynamic by definition.

Now, about the macro usage... Coherence (at least the branch for phoenix 1.2) is not as "dynamic" as it seems. It's not simply a library you import. Let's assume the user's app is named :my_app (top level module will be MyApp). When you run the `mix coherence.install` task, coherence places some files in the application. Those files, which include repo migrations, templates, views and controllers (but not plugs!) are dynamically generated by a mix task that knows the name of your app.

This means that the top level module is MyApp and the code in the views, templates and controllers can import the MyApp.Gettext module instead of the Coherence.Gettext module (it doesn't do that yet, but it can be added). This assumes further that the MyApp.Gettext is the gettext backend for the app -- a sensible convention. I don't know yet if there is a simple way to make this configurable. The macros would then use the locales defined under MyApp's priv/gettext directory.

The plugs are simply imported, and for those I agree with you. You can't use macros with those ones.
For other libraries that have user facing strings but don't generate any files in the app's source tree, I agree with you that functions are the only option.

I hope I didn't make any mistake, and if I did I'm sure @smpallen99 can correct me.

**Comment by josevalim** [comment]
That makes sense, thanks!
-- 


*José Valimwww.plataformatec.com.br
<http://www.plataformatec.com.br/>Founder and Director of R&D*

**Comment by tmbb** [comment]
One other option that wan't discussed above might be to create packages for each translation and have users add then ones they want to their deps. I'm not sure this is the way to go, but I'm just throwing it out there for discussion.

@smpallen99 I think this is the wrong way to do it. This would require publishing a package for each language and force the user to accept the "official" translations. This probably suboptimal... Depending on my users I might prefer certain translations. Let's look at my own language (Portuguese): for older or more linguistically conservative users, I'd translate "email" as "correio electrónico", but for young users I'd leave it as "email". Likewise, some people might prefer a translation for "password" ("palavra passe") and some might prefer the original English word. Another example, you have a template saying something like "oops, something went wrong". For some users, saying "oops" might be a little jarring. They might prefer something like "Unfortunately something might have gone wrong".

The priority should be to ensure the user can customize the translations, and trying to prepackage everything with no possibility for customization strikes me as wrong.

**Comment by tmbb** [comment]
@josevalim Well, thanks a lot for correcting some of my (our?) misconceptions about gettext and for taking the time to help us with this!

**Comment by smpallen99** [comment]
We can't be the only ones doing this?? What do other packages do?

**Comment by smpallen99** [comment]
Ok, I think I finally have a handle on this. I just did some testing with a sample project. All the template stuff works because the Coherence.Web module imports the apps Gettext module.
So we just have to handle the cases for code in the package that isn't generated into the users project. 

I don't quite have a handle on the stuff inside the coherence package yet. I don't get José's comment about having a mix task to copy over the pot files. Perhaps its something to do with compiling the coherence package?

On the topic of po translation files. Could we not generate them, add to priv, and provide a mix task top copy individual languages to the users project?

**Comment by agustif** [comment]
I'm sorry if this is more noise than helpful, but I remembered a couple of blog posts about elixir/phoenix i18n

http://tech.footballaddicts.com/blog/internationalization-of-elixir-applications-with-gettext-and-transifex
http://code.parent.co/practical-i18n-with-phoenix-and-elixir/

**Comment by tmbb** [comment]
José was saying that if you use functions instead of macros you lose the ability of automatically generating the .po files. This means we must provide the user some .po files generated by hand.

**Comment by smpallen99** [comment]
Can't you just setup a little project and generate the po files?

**Comment by smpallen99** [comment]
Thanks @agustif those links were a great read!

**Comment by tmbb** [comment]
> Can't you just setup a little project and generate the .po files?

Not if you use functions instead of macros.
If you use macros then you don't even need the "small project". You can have a Coherence.Gettext backend in coherence's source (as you do already) and import it in the relevant places. Then you can run gettext.extract on coherence's source and generate the .po files.

The catch is that when you generate the coherence.install templates you have to textually replace "import Coherence.Gettext" by "import <%= base %>.Gettext" to use the user's backend

**Comment by tmbb** [comment]
This trick won't help you if you're using functions, but you only need functions in a minority of places, and can do it by hand. There's always the quick'n dirty way of scanning the source code with a regexp to get the strings in that case.

My suggestion for the places where you can't use functions is to define a Coherence.Translate module that defines this function:

```
def dtranslate(text, opts \\ []) do
  Gettext.dgettext Application.get_env(:coherence, :gettext_backend), "coherence", text, opts
end
```

This way the user can provide their own MyApp.Gettext in coherence's config. This can be done by the coherence.install task.

Now you can alias Coherence.Translate and translate any string using `transtate(text)`, which is even less verbose than dgettext

**Comment by smpallen99** [comment]
@tmbb agree on the generated files. Like I said earlier, the views are already done since the Coherence.Web view macro imports the MyProject.Gettext. Not using the name space in the other  mix templates was an over site. 

I'm still confused on the macro vs function approach. Can you elaborate a bit more on this. Thx.

**Comment by tmbb** [comment]
> I'm still confused on the macro vs function approach. Can you elaborate a bit more on this. Thx.

Macros need the gettext backend available at compilation time. We want to use `MyProject.Gettext` as the backend, which is not available at compile time when coherence is compiled as a dependency. This means that when dgettext is used inside a module that is part of `MyProject` (i.e. a module form coherence.install) we can use macros, but inside a coherence module we need to use a function that fetches the `MyProject.Gettext` backend at _runtime_. This means we need a way to supply the backend to the `:coherence` application.

José's suggestion was to add a something like a `:gettext_backend` config option to Coherence that allows the user to supply coherence with the `MyProject.Gettext` backend. The `:gettext_backend` can be fetched at runtime using `Application.get_env(:coherence, :gettext_backend` as I've written above. This sounds like a good idea (and having written `gettext`, José certainly knows what he's talking about :p)

The [gettext docs](https://hexdocs.pm/gettext/Gettext.html) explain the problem well, even if they're not as clear in providing a solution as José was in the above comment.

I might implement the code myself and write a PR but I'm really short on time right now and wouldn't have the time to test it thoroughly (testing this is error prone because it must be done manually and often by visual inspection; just look at the bugs you found in the email templates)

> Not using the name space in the other mix templates was an over site *[assuming you meant "oversight" here]*

But this is something you can trivially fix, right? I'm not 100% on elixir namespaces yet (`use`, `require`, `import`, etc.) but it seems like if all else fails you can just have `import Coherence.Gettext` on the "meta templates" and textually replace it by `import MyProject.Gettext` in the "coherence.install" templates.

I think there should be a section in the gettext docs explaining these problems and explaining José's solution, because as you said we're probably not the ones needing this. When we're done someone should write a PR for the gettext docs explaining what we did.

**Comment by tmbb** [comment]
@smpallen99 Also:

> We can't be the only ones doing this?? What do other packages do?

I've googled it for longer than I should and I've found nothing. I guess not many packages contain user-facing strings like coherence does, or if they do they haven't felt the need to internationalize them properly...

**Comment by smpallen99** [comment]
@tmbb Clear as mud now. But seriously, yes, I'm on board now. Great explanation!!

BTW, use, import and require are pretty simple :)

- `import` brings all the functions directly into your module so they don't need to be name spaced. You can call them without a module name in front of the function name
- `alias` lets you give a shorter name to a module
- `require` is needed it you want to use macros from a module. For example `Logger.info` is  macro so that's why you have to `require Logger` to use it. It does not do an automatic import!
- `use` is a special case. It runs the `__using__/1` macro of the module. It also does an automatic `require` of that module. Thats why you can use macros when you `use` that module. Also, one common practice people do in the `__using__` macro is to import the Module. This way, you can use the functions in the module without a name space.

The `__using__` macro is typically used to inject stuff into the calling module like adding functions. It s kind of like ruby mixins. Note that a module has to implement the `__using__` module in order for it to be `use`d.

**Comment by smpallen99** [comment]
I think what's kind of unique with Coherence is that it mixes generated files, with some of them optional, with other customer facing files are are not part included. Perhaps that's a design flaw??

It makes me ponder on the design. I wonder if there is a better way. Perhaps we should have a generated module (like a view) that contains all messages used by coherence. That way, all messages are in the users app and can be localized there.

**Comment by tmbb** [comment]
@smpallen99 Thanks for the explanation on import, alias, require and use :) I hadn't gotten to that part of the manual

**Comment by tmbb** [comment]
> Perhaps we should have a generated module (like a view) that contains all messages used by coherence. That way, all messages are in the users app and can be localized there.

Can you do this? It doesn't seem possible to use messages from an application module you don't know about at compile time

**Comment by josevalim** [comment]
> Perhaps we should have a generated module (like a view) that contains all messages used by coherence. That way, all messages are in the users app and can be localized there.

I like this idea. If you define a behaviour, the user will also get warnings if they upgrade coherence and are missing some new messages.

If they want to add i18n, then they call something like:

```elixir
import MyApp.Gettext

def welcome_message(opts) do
  gettext "Welcome %{user}", opts
end
```

You don't worry about gettext on coherence and users can still bring it if they need to, all in their apps.

**Comment by tmbb** [comment]
@josevalim Where is the `delcome_message/1` defined? In the user's app?

**Comment by tmbb** [comment]
If so, you need to make sure coherence knows about it. The obvious way is to pass the generated `MyApp.Coherence.Messages` to coherence through a config option. This is probably better than passing the `MyApp.Gettext` backend because it allows you to use macros instead of functions, and you get .po file extraction for free.

Is this what you have in mind?

**Comment by josevalim** [comment]
@tmbb yes, instead of calling the following from coherence:

    def welcome_message(opts) do
      Gettext.gettext some_gettext_backend_from_config(), "Welcome %{user}!", opts
    end

You instead call `messages_module_from_config().welcome_message(opts)`.

Both end-up using options, what changes is the contract (one uses modules+functions, the other imposes gettext).

**Comment by smpallen99** [comment]
There are a couple ways to do this. 

1. Include a default module in coherence. The `messges_module_from_config()` will return this module if the user has not configured an option. Include a generator to create the module in the users project and add config item to define it
2. Always generate the module in the uses project and use it.

Any thoughts on these two alternatives

**Comment by tmbb** [comment]
@josevalim I like this. It's a little more verbose for the user but it's worth it, especially because (1) the module can be automatically generated, thus cutting on typing, and (2) it gives us .po file extraction for free, using the normal gettext workflow and without hacks involving parsing the source with regexps.

@josevalim If this works well would you like PR on gettext docs explaining how to do internationalize a library with user-facing strings using coherence as an example? I think it's appropriate and it's a great opportunity to ensure the elixir ecosystem grows with useful internationalization features by default.

**Comment by tmbb** [comment]
> 2. Always generate the module in the uses project and use it.

This one is simpler (one fewer mix task).

Besides, think it'll be a useful default because it tell the user: "Coherence is configure like this. If you want to change the way the messages are displayed or internationalize your app, just edit this file". It's a friendlier message than: "If you want to change coherence's messages, run this task to generate a file and edit it". This is purely subjective, of course... The price is unnecessary code bloat for someone who doesn't need internationalization or message customization.

**Comment by josevalim** [comment]
Supposedly most people will want to tweak the messages, even if just to change the tone and not necessarily to another language. So I would say you should always generate a module. However, you will need to define a module with the `@callback` for each translation and as well as have default translations for tests, so if you want to ship with one by default, that won't hurt too.

**Comment by smpallen99** [comment]
Thanks @josevalim. That's the approach I'm working on now. Another quick question. I got the impression from your earlier post that you may have been suggesting that the generated file only returned strings (I might be wrong on my impression). I'm thinking that the generated file should include the gettext calls, with an import of the users Gettext module. Thoughts?

**Comment by josevalim** [comment]
@smpallen99 any approach is fine but gettext problably makes more sense, especially if you need to handle pluralization and interpolation.

**Comment by tmbb** [comment]
@smpallen99 If the generated module (under MyProject) includes only strings without the (d)gettext macro calls, how would it help ininternationalizing anything? dgettext would have to be called under the Coherence.Gettext backend (as a function) and we'd be back to where we were before...

I thought the whole point of having a generated module was to be able to call the macros with the MyProject.Gettext backend

**Comment by josevalim** [comment]
@tmbb just to clarify, the generated modules can do whatever it wants. :) it is simply a function contract. The question is about having a good default and using gettext is probably the better default.

**Comment by smpallen99** [comment]
@tmbb No problem. That is what I'm doing. I  just wanted to get José opinion just incase he had an idea I did not think of. 

I ran gettext gettext extract in the Coherence project and I'm using the po file to create the functions. Nice thing about that, it it gives me a reference to the module and line number to add the function call :) I'm about 1/4 though the messages now. 

The approach I'm doing is that messages generated in the Coherence project will use the new Messages module. Messages from the generated code (lives in the users app) will not use the Messages module since they will use the app's Gettext module directly. All messages with use the coherence domain.

**Comment by tmbb** [comment]
@josevalim thanks for the clarification, I agree 100%

**Comment by smpallen99** [comment]
One thing that I have not been able to resolve in my mind is @johnhamelink comment in #173 is how this might fit into his request that we use the a view to render messages for API only support in Coherence. @josevalim any thoughts on this. Also José, I know we are creating a lot of noise for you on this, but given your domain knowledge in this area from your previous life your input is much appreciated.

**Comment by smpallen99** [comment]
BTW, here is what this looks like so far.

```
# behaviour module in Coherence
defmodule Coherence.Messages do

  @callback confirmation_email_sent() :: binary

  def backend do
    Coherence.Config.messages_backend()
  end
end
```

```elixir
     # use the new Messages backend
      send_user_email :confirmation, user, url
      conn
      |> put_flash(:info, Messages.backend().confirmation_email_sent())
```

```elixir
# generated Module
defmodule BlogPhx.Coherence.Messages do
  @behaviour Coherence.Messages

  import BlogPhx.Web.Gettext

  def confirmation_email_send, do: dgettext("coherence", ""Confirmation email sent.")
end
```

**Comment by tmbb** [comment]
@smpallen99 The implementation looks great to me. I don't have very strong opinions on API only functionality or single page applications because I have zero experience in both...

**Comment by smpallen99** [comment]
This is working pretty well. Tests are all passing. Just doing some manual integration testing now. So for, I've only found two issues. The login field isn't translated because it comes from the config. I think I'll leave it as it is and let people change their template if they want.

And the submit button on the new session page. need to track that one down.

**Comment by tmbb** [comment]
> The login field isn't translated

I vote on defaulting to :email (actually `dgettext "coherence", "Email"`) in the template. This way it's internationalized by default. Again, this is purely subjective.

**Comment by tmbb** [comment]
It's basically a tradeoff between minimizing the work of someone who wants to internationalize the :email field and the work of someone who wants to change the login_field AND won't internationalize the app. If they want to internationalize the app they'll have to edit the tenplate anyway.

**Comment by smpallen99** [comment]
I change the default login field way more than I localize my apps that use Coherence. I'm leaving it set to the config item. Plus, it would be worse to display the wrong field than to display the English. Plus, everyone knows Email :)

**Comment by tmbb** [comment]
@smpallen Ok, I bow before the experience of someone whi has made more than one app using Coherence :)

**Comment by smpallen99** [comment]
So, just to be clear. 

We are NOT going to provide pre-translated `.po` files with Coherence. I don't believe we can add much value with canned translations. The way we have set this up, all Coherence messages now use the users app's Gettext module with the "coherence" domain. Since a site is going to require translations for the main site, it only makes sense IMHO to translate the Coherence messages at the same time. Furthermore, this provides the opportunity to tweak the messages before they are sent off to be translated.

**Comment by tmbb** [comment]
Before you close this issue, a meta point (not directly related to coherence):

1. We have successfully used gettext to internationalize the coherence package, a library with user facing strings in a way that give the user complete control on how to extract and generate the tarnslations.

2. The approach is elegant and simple but non obvious: it mixes textually generated code with supplying behaviors at config time. This is weird but works really well. We'd probably take ages to converge on this design it if it weren't for José's help, and I think we should make it easier for others to find this information.

3. The obvious place to document this is in gettext's hexdocs. I'd like to submit a pull request to gettext describing the approach we took as a way of providing a working recipe for other teams. @smpallen99 do you agree with this?

**Comment by smpallen99** [comment]
@tmbb I would start by raising an issue against Gettext and see what reaction you get from that. Other alternatives:

- A blog post
- A topic in Elixir Forum

The nice thing about a blog post, is that you can submit it to Elixir Radar (I think thats what its called). That way it goes out to a wide audience. I'll also tweet it. My tweets usually get retweeted by several that have a huge following.

**Comment by tmbb** [comment]
Ok, it's not my top priority now but I'll try to find time for it

**Comment by tmbb** [comment]
Mark this as closed, maybe?
